scylla/src/transport/load_balancing/default.rs

```diff
@@ -283,13 +309,19 @@ impl DefaultPolicy {
         predicate: impl Fn(&NodeRef<'a>) -> bool,
         cluster: &'a ClusterData,
     ) -> impl Iterator<Item = NodeRef<'a>> {
+        let predicate = move |node| match replica_location {
+            ReplicaLocation::Any | ReplicaLocation::Datacenter => predicate(&node),
+            ReplicaLocation::DatacenterAndRack => {
+                predicate(&node) && node.rack == self.preferred_rack
+            }
+        };
         let should_be_local = match replica_location {
             ReplicaLocation::Any => false,
             ReplicaLocation::Datacenter | ReplicaLocation::DatacenterAndRack => true,
         };
         self.nonfiltered_replica_set(ts, should_be_local, cluster)
             .into_iter()
-            .filter(predicate)
+            .filter(move |node: &NodeRef<'a>| predicate(node))
```

Comment:
This change should also be applied to the previous commit - it does not compile without it.